subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ExecutionStateChanges.java

```diff
@@ -16,32 +16,26 @@
 
 package org.gradle.internal.execution.history.changes;
 
+import com.google.common.collect.ImmutableSortedMap;
 import org.gradle.internal.change.Change;
 import org.gradle.internal.change.ChangeVisitor;
-import org.gradle.internal.execution.history.AfterPreviousExecutionState;
-
-import java.util.Optional;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
 
 /**
  * Represents the complete changes in execution state
  */
 public interface ExecutionStateChanges {
 
-    int MAX_OUT_OF_DATE_MESSAGES = 3;
-
-    /**
-     * Returns changes to input files only, or {@link Optional#empty()} if a full rebuild is required.
-     */
-    Optional<Iterable<Change>> getInputFilesChanges();
-
     /**
      * Visits any change to inputs or outputs.
      */
     void visitAllChanges(ChangeVisitor visitor);
 
-    /**
-     * The base execution the changes are calculated against.
-     */
-    // TODO Use AfterPreviousExecutionState from context instead
-    AfterPreviousExecutionState getPreviousExecution();
+    <T> T visitInputFileChanges(IncrementalInputsVisitor<T> visitor);
+
+    interface IncrementalInputsVisitor<T> {
+        T visitRebuild(ImmutableSortedMap<String, CurrentFileCollectionFingerprint> allFileInputs);
+
+        T visitIncrementalChange(Iterable<Change> inputFileChanges);
```

Comment:
There shouldn't be two methods here. Instead, we should pass an `Iterable<Change>` always. The implementation of that iterable can be lazy to avoid eager evaluation in the rebuild case when the value doesn't actually get used.

I think the method name should be plural, too, i.e. `visitIncrementalChanges()`.

Reply:
If we pass an `Iterable<Change>`, how does the task action know if it is executed incrementally or not? How about we have an `InputFileChanges` object, which can be queried if the execution is incremental and about the file changes?

Reply:
Ah, right, I thought this through once already. Yeah, let's pass _the_ `InputFileChanges` (or better yet, `InputChanges`) object here. Or even better: handle all of this in `:execution`, and just wrap `IncrementalTaskInputs` around it.

Reply:
I did this in the PR https://github.com/gradle/gradle/pull/8746 which also introduces the `InputChanges` object for tasks and transforms: https://github.com/gradle/gradle/blob/7267491f737b9f83ea7492b7887fbf1450384b42/subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ExecutionStateChanges.java#L21-L32